### ISSUE TITLE
[Feature] Semantic highlight support for C++?

### ISSUE THREAD

**Issue by ronakg** [sainnhe/gruvbox-material-vscode#228]
- Operating system/version: macOS Monterey
- VSCode version: 

### Settings

NA

### Steps to reproduce this bug

NA

### Actual behaviour

NA

### Expected behaviour

This is a request for semantic highlighting support for C++. Right now the theme doesn't define any special semantic tokens for C++, so the VSCode theme looks inferior to the Vim version.

**Comment by sainnhe** [comment]
![Screenshot_20220611_120417](https://user-images.githubusercontent.com/37491630/173171957-8a6fe96b-5146-4ac7-b8b9-b332a0ad6ab9.png)

IMO, it works fine for me. Do you think there is anything thing need to be optimized?

---

Btw, you can follow the [CONTRIBUTING](https://github.com/sainnhe/gruvbox-material-vscode/blob/master/CONTRIBUTING.md) guide to modify the source code to what you want.

**Comment by ronakg** [comment]
This is just the default C++ syntax highlighting. The theme doesn't highlight semantic tokens like `property`, `namespace`, `parameter`, etc. which are really useful for reading the code efficiently. 

Ideally the theme should have highlighting for all the tokens documented here - https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#standard-token-types-and-modifiers

**Comment by ronakg** [comment]
Here's the semantic token colors defined by the Mayukai theme for reference - https://github.com/GulajavaMinistudio/Mayukai-Theme/blob/master/themes/Mayukai-mirage-color-theme-semantic.json#L198

**Comment by sainnhe** [comment]
> The theme doesn't highlight semantic tokens like property, namespace, parameter, etc.

It's intentional, variables are supposed to be white. This design is also used in neovim.

> Ideally the theme should have highlighting for all the tokens documented here -

I don't think so. We only need to use semantic highlighting engine when the default highlighting engine cannot highlight a scope.

**Comment by sainnhe** [comment]
> I think the idea behind semantic highligting is that we want to highlight things based on their semantic nature. A parameter is different from a local variable, which is different from a class member. So they all should be highlighted in a different (maybe just minor) manner. I'd even argue that properties of a variable like readonly or not is worth taking into consideration.

First of all, it is by design. I know variables, parameters and properties are different things, but in my design they are not supposed to be highlighted. I think we should not highlight everything, because if everything is special, then nothing is special.

Another reason is that there aren't enough colors to get all them highlighted. There are 7 colors in total, and they have been used in things like keywords, operators, etc. If we highlight all of them, the colors will be conflicted.

> I feel that if Semantic tokens are available, then those should be used for highlighting as they are a lot more accurate compared to regex based tokens.

Honestly, semantic highlighting engine is not as good as you think. I personally think that all semantic highlighting in all language servers except rust analyzer are defective, only very limited highlighting can be done. I think the correct way to use it is to use it as a complement to regex highlighting, and only use semantic highlighting when regex highlighting can't do what we want.

**Comment by ronakg** [comment]
Fair enough. Can you help me customize the highlighting based on semantic tokens for `paramters`, `property` and `namespace`?

**Comment by sainnhe** [comment]
Launch command palette, search for "Inspect tokens", this will pop up a floating window to show the tokens under current cursor.

As for customizing settings.json, sorry but I haven't tried it yet, you can search for a blog to find how to do it.
